### DISCUSSION TITLE
Make BASS components less `internal`

### DISCUSSION THREAD

**Discussion by Feodor0090** [ppy/osu-framework#6112] [Q&A]
Primary usage scenario for o!f audio system is obtaining tracks from `ResourceStore` or internet via `ITrackStore` by name.

My use case is quite more complex - i'm getting tracks data as `byte[]` from another application, i don't know arrived track format and it maybe an SMF file (midi). I ended up using BASSMIDI extension and created this thing: https://gist.github.com/Feodor0090/7e9f9a1d53aeb5445a5fabce03751af2

The question is in reflection usage. `TrackBass` constructor is internal, so it can't be created directly. Also, to interact with BASSMIDI i need track handle, but it's exposed via `IBassAudioChannel` that is also internal.

What were the reasons not making these public, and can they be made public to make BASS plugins integration with o!f easier?

**Comment by bdach** [comment]
They are internal because BASS is - in the large scale of framework things - an implementation detail. Consumers should not care what's underneath.

MIDI playback support is mostly out of scope. It's either what you have or forking framework.

**Comment by peppy** [comment]
`TrackBass` used to be public but from my recollection we intentionally changed to `internal` to avoid external usage (because we want to be able to change the API).

I'd say for now you're best off forking of copy-pasting classes you need into your local project. We're severely manpower limited and not looking to increase scope for now.
